Variable.bind newer

chenyuxyz opened the pull request:
update based on #1962 making a new draft. (1) added Variable.bind, (2) removed var_vals from LazyBuffer, (3) implemented ShapeTracker.unbind and used in jit check

chenyuxyz commented:
this is ready for review. I will do a followup to pass bound Variable in jit and remove the jit_ctx.

the method cache thing is more complicated than I thought, will attempt separately

geohot commented on tinygrad/shape/view.py:
```diff
@@ -20,19 +20,29 @@ class View:
 
+  def unbind(self) -> View:
+    variables:List[Variable] = []
+    for x in self.shape+self.strides+(self.offset,):
```
This should probably be reused code with st.var_vals

chenyuxyz replied:
refactored some vars logic

geohot commented on tinygrad/shape/view.py:
```diff
@@ -20,19 +20,29 @@ class View:
   strides:Tuple[sint, ...]
   offset:sint
-  mask:Optional[Tuple[Tuple[sint, sint], ...]]
+  mask:Optional[Tuple[Tuple[int, int], ...]]
```
Are you sure this is okay? How does cat work?

chenyuxyz replied:
no that was incorrect. fixed and adding mask as a source of variable. I think it's possible to prove that only check a subset of (shape, strides, offset, mask) suffices, but checking all 4 is easier to reason

geohot commented on tinygrad/ops.py:
```diff
@@ -214,7 +214,7 @@ def __call__(self, rawbufs:List[RawBuffer], var_vals:Optional[Dict[Variable, int
   def __init__(self, buffer: Type[RawBuffer], linearizer_opts, renderer, runtime, synchronize=lambda: None, batch_exec=BasicBatchExecutor):
     self.buffer, self.linearizer_opts, self.renderer, self.runtime, self.synchronize, self.batch_exec = buffer, linearizer_opts, renderer, runtime, synchronize, batch_exec
-    self.method_cache: Dict[Any, ASTRunner] = {}
+    self.method_cache: Dict[LazyOp, ASTRunner] = {}
```
👍

geohot commented:
I don't see var bufferops unbinding the shapetracker. No bound variables should ever be in the AST that goes into the linearizer, can you check this?

ZenginU commented on tinygrad/shape/view.py:
```diff
@@ -33,7 +33,9 @@ def create(shape:Tuple[sint, ...], strides:Optional[Tuple[sint, ...]]=None, offs
-  def vars(self) -> List[Variable]: return dedup(functools.reduce(operator.add, [x.vars() for x in self.shape+self.strides+(self.offset,) if isinstance(x, Node)], []))
+  def vars(self) -> List[Variable]:
+    flatten_mask = tuple(flatten(self.mask)) if self.mask is not None else tuple()
+    return dedup(functools.reduce(operator.add, [x.vars() for x in self.shape+self.strides+(self.offset,)+flatten_mask if isinstance(x, Node)], []))
```
Maybe use flatten here as well instead of functools.reduce ?

chenyuxyz replied:
flatten does like tuple because it's not Iterator, I also revert the use of flatten in flatten_mask. There's also itertools.chain, but I think these are close enough in terms of speed and readability

chenyuxyz commented:
added the unbind in op shapetracker and verified method cache works now since all variables are free. A side note is that we also need to unbind the reduce op arg if it contains variables. We only have (Variable, NumNode, MulNode, SumNode) in symbolic shape, so I added unbind for MulNode and SumNode.

geohot commented on tinygrad/shape/symbolic.py:
```diff
@@ -194,6 +205,7 @@ def __mod__(self, b: Union[Node, int]):
   def get_bounds(self) -> Tuple[int, int]:
     return (self.a.min*self.b, self.a.max*self.b) if self.b >= 0 else (self.a.max*self.b, self.a.min*self.b)
   def substitute(self, var_vals: Dict[VariableOrNum, Node]) -> Node: return self.a.substitute(var_vals) * (self.b if isinstance(self.b, int) else self.b.substitute(var_vals))
+  def unbind(self) -> Tuple[Node, None]: return self.substitute({v: v.unbind()[0] for v in self.vars() if v.val is not None}), None
```
Can this just be on the root class? I see it's copy pasted below

chenyuxyz replied:
done

geohot commented on tinygrad/lazy.py:
```diff
@@ -79,29 +79,28 @@ def get_single_root(root:LazyBuffer) -> LazyBuffer: return get_single_root(cast(
 def get_movementroot_contiguous(x:LazyBuffer) -> LazyBuffer: return get_movementroot_contiguous(cast(LazyBuffer, x.op.src[0])) if not x.realized and x.op.op == LoadOps.CONTIGUOUS else (get_movementroot(x, True) if x.optype == MovementOps and x.st.contiguous else x)
 
-def var_vals_from_ast(ast:LazyOp) -> List[Variable]: return dedup(functools.reduce(operator.add, [x.arg.st.var_vals() for x in ast.get_lazyops() if x.op in BufferOps], []))
+def vars_from_ast(ast:LazyOp) -> List[Variable]: return dedup(functools.reduce(operator.add, [x.arg.st.vars() for x in ast.get_lazyops() if x.op in BufferOps], []))
```
As you pointed out, what about vars in ReduceOps?

geohot commented on tinygrad/lazy.py:
```diff
@@ -31,7 +31,7 @@ def _ast_reduceops(op:LazyOp) -> LazyOp:
     assert isinstance(src.op, LazyOp), "if not src.realized, then src.op must be a LazyOp"
     if MERGE_ELEMENTWISE_INTO_REDUCE and src.optype is BinaryOps and len(src.children) <= 1: src = src.op
-  return LazyOp(op.op, (src,), op.arg)
+  return LazyOp(op.op, (src,), tuple(a.unbind()[0] if isinstance(a, (Variable, MulNode, SumNode)) else a for a in op.arg) if op.arg is not None else None)
```
Should this be here? Why ever allow bound variables in ReduceOps?

chenyuxyz replied:
I will go through this PR tomorrow again to make sure everything is good, but I have some clarifying questions about your comments.

so I moved the unbind reduceop.arg logic into the `r()` function, and I think the `arg` of a reduceop can only be `new_shape`. So it cannot be None and cannot be tuples either. I updated the logic to unbind that new_shape before creating LazyOp, and keep the bound st in LazyBuffer.

since reduce arg is new_shape, which is a subset of the original shape, it should suffice to get variables from the shapetracker.

geohot replied:
Oh sorry, I was thinking movementops. ReduceOp can't be none or tuples. I think the unbind in the _reduce_ops function you have now is correct.

chenyuxyz commented:
rebased to master.

so we can have bound variables in lazybuffer shapetracker and movementops, and no bound variables (unbind in bufferops creation and reduceop arg, no need for elementwise ops) in AST going into linearizer. And we don't need to change `Variable.substitute` since all symbolic are unbound after compile time.

also checked with DEBUG=4 on various symbolic op tests and tested with llama. I think this is good to go
